### PR TITLE
feat: upgrade dependencies to support Apollo Server 3 and latest Apollo Gateway

### PR DESCRIPTION
**Pull request by sjungling**
### Problem
When attempting to use with a combination of Apollo Server 3.3.x and Apollo Gateway 0.41.x I was running into issues satisfying dependencies version without the use of `--force` installation (NPM 7). 

### Solution
Upgrade all dependencies in range (except TypeScript), update tests to conform to Apollo Server 3 config, and use types from Apollo for `process` and `willSendRequest`. Also updated some grammatical nuance on the `README` and updated the example code.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -19,7 +19,8 @@
     "strictPropertyInitialization": false,
     "declaration": true,
     "baseUrl": ".",
-    "outDir": "./build"
+    "outDir": "./build",
+    "skipLibCheck": true
```
Comment (sjungling) [inline tsconfig.json:23]: Using TypeScript 4.4.x without this setting there were some excessive type errors being thrown out from `node_modules`

File: test/gen-service.ts [hunk 2]
```diff
@@ -135,9 +135,8 @@ const genService = (
           typeDefs,
         },
       ]),
-      subscriptions: false,
-      uploads: false,
```
Comment (sjungling) [inline test/gen-service.ts:139]: Neither of these options are apart of Apollo Sever 3

**Comment by cabelitos** [comment]
@sjungling thank you very much for the PR.  Could you please take a look in my comments?

File: lib/FileUploadDataSource.ts [hunk 3]
```diff
@@ -215,7 +215,7 @@ export default class FileUploadDataSource extends RemoteGraphQLDataSource {
     const options = {
       ...request.http,
       // Apollo types are not up-to-date, make TS happy
-      body: (form as unknown) as string,
+      body: form as unknown as string,
```
Comment (cabelitos) [inline lib/FileUploadDataSource.ts:218]: please avoid changing styling. Did eslint complained?

File: lib/FormData.ts [hunk 4]
```diff
@@ -11,14 +11,14 @@ export default class FormData extends BaseFormData {
...
+        callback(null, null as unknown as number);
       }
     };
     super.getLength(cb);
   }
 
   getLengthSync(): number {
     const len = super.getLengthSync();
-    return Number.isNaN(len) ? ((null as unknown) as number) : len;
+    return Number.isNaN(len) ? (null as unknown as number) : len;
```
Comment (cabelitos) [inline lib/FormData.ts:22]: ditto, no need to change this.

Comment (cabelitos) [reply lib/FormData.ts:22]: If this is related to eslint complains, please put it a different commit.

File: tsconfig.json (hunk 1 above)
Comment (cabelitos) [reply tsconfig.json:23]: I don't think that this is a good idea. what's the typescript complain?

File: lib/FormData.ts (hunk 4 above)
Comment (sjungling) [reply lib/FormData.ts:22]: @cabelitos yes, it was a prettier complaint surfaced through eslint.
If you want, i could disable that particular check for that line and in `FileUploadDataSource` via 
```js
// eslint-disable-next-line prettier/prettier
```

File: tsconfig.json (hunk 1 above)
Comment (sjungling) [reply tsconfig.json:23]:
```shell
$ tsc --noEmit
node_modules/@types/express-serve-static-core/index.d.ts:30:10 - error TS2305: Module '"qs"' has no exported member 'ParsedQs'.

30 import { ParsedQs } from 'qs';
            ~~~~~~~~

node_modules/@types/express/index.d.ts:58:29 - error TS2694: Namespace 'serveStatic' has no exported member 'RequestHandlerConstructor'.

58     var static: serveStatic.RequestHandlerConstructor<Response>;
                               ~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@types/serve-static/node_modules/@types/express-serve-static-core/index.d.ts:501:18 - error TS2430: Interface 'Response<ResBody>' incorrectly extends interface 'ServerResponse'.
  Types of property 'req' are incompatible.
    Type 'Request<ParamsDictionary, any, any, Query> | undefined' is not assignable to type 'IncomingMessage'.
      Type 'undefined' is not assignable to type 'IncomingMessage'.

501 export interface Response<ResBody = any> extends http.ServerResponse, Express.Response {
                     ~~~~~~~~


Found 3 errors.
```

these appear to be transitive dependencies of the `apollo-server-express` dependency

File: lib/FileUploadDataSource.ts (hunk 3 above)
Comment (sjungling) [reply lib/FileUploadDataSource.ts:218]: yes, I can revert and disable that particular line from eslint if you prefer.

**Comment by lassesteffen** [comment]
Can we get this merged?

**Comment by fredleger** [comment]
We desperately need this! Love luck and profit to the one that merge it 😂

**Comment by cabelitos** [comment]
hey guys, sorry for the *long* delay. I'll take a look.

**Comment by lassesteffen** [comment]
With the latest versions of graphql gateway it suddenly stopped worked for me even with this adapted version. Maybe you can check that as well!

**Comment by cabelitos** [comment]
Hey guys, I close this commit because due to conflicts. However it's included in here ( I fixed the conflicts). https://github.com/profusion/apollo-federation-file-upload/pull/42

I also took the liberty to update all the deps once again, since they are now a little bit outdated.

I bumped all the peers deps to

```
    "@apollo/gateway": "^0.46.0",
    "apollo-server-env": "^4.2.1",
    "apollo-server-types": "^3.5.1",
    "graphql-upload": "^13.0.0"
```


Please use version 3.0.0 which is now available on NPM https://www.npmjs.com/package/@profusion/apollo-federation-upload

Sorry for the delay and thank you for your patience

cc:
@sjungling @lassesteffen @fredleger

**Pull request closed without merging**